### ISSUE TITLE
Proposal: Change keyword for the Omnibox API. 

### ISSUE THREAD

**Issue by newRoland** [hober/web-extensions#545]
The Omnibox API offers no way to rename the keyword. The user might not like the default keyword, but would still want to use the extension. The user might also want to use more than one extension that use the same keyword. 

This capability can be implemented as a developer API, a user-configurable option offered by the browser, or both. 

## API Proposal
```js 
browser.omnibox.setKeyword("foo")  // set to foo. 

browser.omnibox.setKeyword(null)  // set to empty string or null to disable omnibox. 
```

## Browser Option Proposal

As a user configurable setting through the browser. Initially suggested by @tophf

**Comment by tophf** [comment]
...and the callback/Promise would return an error in case the shortcut is invalid or already occupied.

The browsers could also allow editing it in their extension shortcuts UI because an omnibox shortcut is conceptually a variant of a keyboard shortcut to activate an extension.

**Comment by xeenon** [comment]
Yes, nothing prevents browsers from letting the user edit this in the browser's settings UI.

**Comment by newRoland** [comment]
@tophf @xeenon That's also fine. Currently, I don't think any browser allows doing that. 

For Safari and Firefox, I see no options for renaming, disabling a extension's shortcut. 

For Chromium, in chrome://settings/searchEngines, it does show an option to "disable" a shortcut, but that just disables the extension. 

One advantage to also having an API would be to implement all the extension's settings in one place.

**Comment by bershanskiy** [comment]
I like the idea of letting user explicitly change the shortcut via browsers' native UI.

On the other hand, the API method like `browser.omnibox.setShortcut()` seems ripe for abuse by deceptive extensions which could attempt to impersonate other extensions. Perhaps, this API should have a user activation requirement and an explicit acknowledgement by the user (e.g., a permission prompt asking "Do you want to change shortcut of <extension_name> from <old_shortcut> to <new_shortcut>?")

**Comment by newRoland** [comment]
As @tophf suggested, it should return an error if the shortcut is already occupied. With that in place, I don't see enough room for abuse to warrant a permission dialogue.

**Comment by bershanskiy** [comment]
Would this kind of UI work?
![image](https://github.com/w3c/webextensions/assets/45960703/aee49c6e-5ccd-49ba-8397-ec017e6a8bf3)

**Comment by newRoland** [comment]
That looks great! One little change I would make is replacing "Disable extension" with "Disable keyword". If the user wants, they could disable/delete the extension through Manage. 

I didn't consider Disable suggestions before, but that would also be useful.

**Comment by bershanskiy** [comment]
@newRoland I'm still not entirely sure what exactly you are proposing. Could you please clarify?

The relevant code can be seen here:
 - [Markup is in `omnibox_extension_entry.html`](https://github.com/chromium/chromium/blob/main/chrome/browser/resources/settings/search_engines_page/omnibox_extension_entry.html)
 - [Script is in `omnibox_extension_entry.ts`](https://github.com/chromium/chromium/blob/main/chrome/browser/resources/settings/search_engines_page/omnibox_extension_entry.ts)

The current UI render is:
![image](https://github.com/w3c/webextensions/assets/45960703/c1691360-c2fe-40ea-b2f5-bbdb464904b5)

The current entries are:
 - "Manage" - opens extension's management page at `chrome://extensions/?id=<extension_id>`
 - "Disable" - disables the entire extension

Could you please write down the desired entries and desired behaviors? Preferably, mentioning the exact button label and exact behavior.

Also, the desired UI appearance for the edit thingy. Here are a few considerations (see [video for of UI used for other edit menus](https://github.com/w3c/webextensions/assets/45960703/bb0831b0-9112-43cf-ad83-14cfb1dbb5b3)):
 - Should keyword edit action be initiated by a button in a three-dot-menu or by a "pencil" button?
 - Should the edit UI be an inline input field or an input field on a popover which grabs focus of the current page?

**Comment by newRoland** [comment]
No preference in terms of UI. As long as the user can edit and disable the keyword, that would be great.

**Comment by bershanskiy** [comment]
I filed a [Chromium bug](https://issues.chromium.org/issues/325619961) to track this feature.

**Comment by bershanskiy** [comment]
I just remembered that the manifest entry for the omnibox shortcut is `omnibok.keyword` ([Chrome docs](https://developer.chrome.com/docs/extensions/reference/api/omnibox#manifest), [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/omnibox)), so perhaps the method name `browser.omnibox.setKeyword()` might be more appropriate. (I'm still unsure whether the JS API makes sense.)

**Comment by oliverdunk** [comment]
Chrome is supportive of this API. It would need to go through security and privacy review, but we don't expect any significant trouble there.

A small thing to consider would be if we should change the `keyword` key in the manifest to `default_keyword`, to indicate that it may change.

**Comment by Rob--W** [comment]
A bug report has been opened in Firefox, at https://bugzilla.mozilla.org/show_bug.cgi?id=1948615
